unicorn/include/cocaine/unicorn.hpp

```diff
@@ -0,0 +1,219 @@
+
+#include "cocaine/idl/unicorn.hpp"
+#include "cocaine/unicorn/path.hpp"
+#include "cocaine/unicorn/value.hpp"
+
+#include "cocaine/zookeeper/connection.hpp"
+#include "cocaine/zookeeper/session.hpp"
+
+#include <cocaine/api/service.hpp>
+#include <cocaine/rpc/dispatch.hpp>
+
+#include <asio/ip/tcp.hpp>
+#include <asio/deadline_timer.hpp>
+
+
+using namespace cocaine::unicorn;
+namespace cocaine { namespace service {
+
+class unicorn_service_t:
+    public api::service_t,
+    public dispatch<io::unicorn_tag>
+{
+public:
+    virtual
+    const io::basic_dispatch_t&
+    prototype() const;
+
+    unicorn_service_t(context_t& context, asio::io_service& asio, const std::string& name, const dynamic_t& args);
+    friend class unicorn_dispatch_t;
+    friend class distributed_lock_t;
+    friend void release_lock(unicorn_service_t* service, const path_t& path);
+    const std::string& get_name() const { return name; }
+private:
+    std::string name;
+    zookeeper::session_t zk_session;
+    zookeeper::connection_t zk;
+    std::shared_ptr<cocaine::logging::log_t> log;
+};
+
+/**
+* Service for providing access to unified configuration service.
+* Currently we use zookeeper as backend.
+* For protocol methods description see include/cocaine/idl/unicorn.hpp
+*/
+class unicorn_dispatch_t :
+    public dispatch<io::unicorn_final_tag>
+{
+public:
+
+    typedef std::shared_ptr<zookeeper::handler_scope_t> scope_ptr;
+    /**
+    * Typedefs for result type. Actual result types are in include/cocaine/idl/unicorn.hpp
+    */
+    struct response {
+        typedef deferred<result_of<io::unicorn::put>::type> put;
+        typedef deferred<result_of<io::unicorn::create>::type> create;
+        typedef deferred<result_of<io::unicorn::del>::type> del;
+        typedef deferred<result_of<io::unicorn::increment>::type> increment;
+        typedef streamed<result_of<io::unicorn::subscribe>::type> subscribe;
+        typedef streamed<result_of<io::unicorn::lsubscribe>::type> lsubscribe;
+        typedef deferred<result_of<io::unicorn::lock>::type> lock;
+    };
+
+    unicorn_dispatch_t(const std::string& name, unicorn_service_t* parent);
+
+    response::put
+    put(path_t path, value_t value, version_t version);
+
+    response::create
+    create(path_t path, value_t value);
+
+    response::del
+    del(path_t path, version_t version);
+
+    response::subscribe
+    subscribe(path_t path);
+
+    response::lsubscribe
+    lsubscribe(path_t path);
+
+    response::increment
+    increment(path_t path, value_t value);
+
+    /**
+    * Callbacks to handle async ZK responses
+    */
+    struct nonode_action_t;
+
+    struct subscribe_action_t;
+
+    struct lsubscribe_action_t;
+
+    struct put_action_t;
+
+    struct create_action_base_t;
+    struct create_action_t;
+
+    struct del_action_t;
+
+    struct increment_action_t;
+    struct increment_create_action_t;
+
+private:
+    scope_ptr handler_scope;
+    unicorn_service_t* service;
```

Comment:
unicorn service is created on startup and exists till destructor of context_t. Also context_t in destructor first destroys pool of execution_units which destroys session and sessions destroy dispatch. Only after that service is destroyed.